### ISSUE TITLE
SPM doesn't enumerate source files properly in different configurations

### ISSUE THREAD

**Issue by compnerd** [apple/swift-package-manager#7044]
### Description

When building https://github.com/thebrowsercompany/swift-composable-architecture in release mode, the generated OFM for OpenCombine is invalid, e.g.

```
Building for production...
<unknown>:0: error: supplementary output file map 'C:\\Users\\abdulras\\AppData\\Local\\Temp\\TemporaryDirectory.T3w0D9\\supplementaryOutputs-1' is missing an entry for 'S:\\SourceCache\\thebrowsercompany\\swift-composable-architecture\\.build\\checkouts\\OpenCombine\\Sources\\OpenCombine\\AnyCancellable.swift' (this likely indicates a compiler issue; please submit a bug report (https://swift.org/contributing/#reporting-bugs))
```

Note that building OpenCombine directly does not exhibit this issue.

### Expected behavior

The internal generated state would be correct.

### Actual behavior

The internal generated state is incorrect.

### Steps to reproduce

1. git clone https://github.com/thebrowsercompany/swift-composable-architecture
2. cd swift-composable-architecture
3. git checkout develop
4. swift build -c release

### Swift Package Manager version/commit hash

Swift Package Manager - Swift 5.10.0-dev

### Swift & OS version (output of `swift --version ; uname -a`)

Swift version 5.11-dev (LLVM 61776217335130e, Swift 4bcc25e99094074)
Target: x86_64-unknown-windows-msvc

**Comment by neonichu** [comment]
I don't believe SwiftPM creates supplementary output file maps but the driver does.

@shahmishal could we move this issue to the driver repo?
